car-pool/car-mirror/SPEC.md

```diff
@@ -201,151 +167,27 @@ The Provider MAY garbage collect its session state when it has exhausted its gra
-    ▼         ▼           │                                  │              ▼          ▼
-┌─────────────────────┐   │                                  │         ┌────────────────────┐
-│                     │   │                                  │         │                    │
-│      Updated        │   │                                  │         │       Updated      │
-│  Bloom & CID Roots  │   │                                  │         │      Hash Set      │
-│                     │   │                                  │         │                    │
-└─────────────────────┘   │                                  │         └────────────────────┘
-```
+![](./diagrams/pull.svg)
 
 ## 3.3 Push Protocol 📤
 
 ### 3.3.1 High Level
 
 The push protocol is very similar to pull (§3.2), roughly inverted. The major difference aside from direction of data flow is that the first round MAY be a "cold call". This prioritizes starting the first round with zero information about the Provider, under the assumption that the first branch of the graph will either not contain many duplicate blocks, or if it does that it is doing as well as the prior art CAR transfers. This strategy prioritizes limiting round trips over reducing bandwidth.
 
-The sending Requestor begins with a local phase estimating what the Provider has in its store. This may be from stateful information (e.g. Bloom filters and CIDs) learned in a previous round, or by using a heuristic such as knowing that the provider has a previous copy associated with an IPNS record or DNSLink. If no information is available, the estimate is the empty set.
+The sending Requestor begins with a local phase estimating what the Provider has in its store. This estimate is called the Provider Graph Estimate. This may be from stateful information (e.g. Bloom filters and CIDs) learned in a previous round, or by using a heuristic such as knowing that the provider has a previous copy associated with an IPNS record or DNSLink. If no information is available, the estimate is the empty set.
 
 The Requestor performs graph traversal of the data under the CID to push, appending blocks to a CAR file. This CAR MAY be discrete or streaming, depending on the transport. All other factors being equal, breadth-first traversal is RECOMMENDED. Since the Provider is not expected to use the data immediately, it exposes the largest number of roots, and thus grants the highest chance of discovering shared subgraphs. An implementation MAY choose a different traversal strategy, for example if it knows more about the use.
 
+On a partial cold call, the Provider Graph Estimate MUST contain the entire graph minus CIDs in the initial payload. The Provider MUST respond with a Bloom filter of all CIDs that match the Provider Graph Estimate, which is called the Provider Graph Confirmation. On subsequent rounds, the Provider Graph Estimate continues to be refined until is is empty or the entire graph has been synchronized.
+
```

Comment:
The actual substantive change

Reply:
Those are helpful additions and I believe I'm tracking now.  If I can restate to confirm...

It's as if in the first round we just send some number of CIDs in a CAR file along with a bloom filter that says "we're not sending any more to you because we've guessed that you already have everything we planned on pushing".  But the provider uses that provider graph estimate as a way of learning what the requestor planned on pushing  (walking non-terminating leaf CIDs received in the CAR and possibly `diff` as well, to generate CIDs and comparing with the requestor graph estimate bloom.  The provider returns a provider graph confirmation bloom, which is really just a new bloom that says what the provider has from what the provider thinks the requestor plans on pushing.

Reply:
```suggestion
The sending Requestor begins with a local phase estimating what the Provider has in its store. This estimate is called the Provider Graph Estimate. This may be from stateful information (e.g. Bloom filters and CIDs) learned in a previous round, or by using a heuristic such as knowing that the provider has a previous copy associated with an IPNS record or DNSLink. If no information is available, the estimate is the empty set.

The Requestor performs graph traversal of the data under the CID to push, appending blocks to a CAR file. This CAR MAY be discrete or streaming, depending on the transport. All other factors being equal, breadth-first traversal is RECOMMENDED. Since the Provider is not expected to use the data immediately, it exposes the largest number of roots, and thus grants the highest chance of discovering shared subgraphs. An implementation MAY choose a different traversal strategy, for example if it knows more about the use.

On a partial cold call, the Provider Graph Estimate MUST contain the entire graph minus CIDs in the initial payload. The Provider MUST respond with a Bloom filter of all CIDs that match the Provider Graph Estimate, which is called the Provider Graph Confirmation. On subsequent rounds, the Provider Graph Estimate continues to be refined until it is empty or the entire graph has been synchronized.

```

Reply:
That does not make clear what I changed. 😬 

"to be refined until is is" -> "to be refined until it is"

Reply:
```suggestion
The sending Requestor begins with a local phase estimating what the Provider has in its store. This estimate is called the Provider Graph Estimate. This may be from stateful information (e.g. Bloom filters and CIDs) learned in a previous round, or by using a heuristic such as knowing that the provider has a previous copy associated with an IPNS record or DNSLink. If no information is available, the estimate is the empty set.

The Requestor performs graph traversal of the data under the CID to push, appending blocks to a CAR file. This CAR MAY be discrete or streaming, depending on the transport. All other factors being equal, breadth-first traversal is RECOMMENDED. Since the Provider is not expected to use the data immediately, it exposes the largest number of roots, and thus grants the highest chance of discovering shared subgraphs. An implementation MAY choose a different traversal strategy, for example if it knows more about the use.

On a partial cold call, the Provider Graph Estimate MUST contain the entire graph minus CIDs in the initial payload. The Provider MUST respond with a Bloom filter of all CIDs that match the Provider Graph Estimate, which is called the Provider Graph Confirmation. On subsequent rounds, the Provider Graph Estimate continues to be refined until it is empty or the entire graph has been synchronized.

```

That does not make clear what I changed. 😬 

"to be refined until is is" -> "to be refined until it is"